mirrord/agent/src/steal/http/filter.rs

```diff
@@ -35,6 +37,108 @@ const H2_PREFACE: &[u8; 14] = b"PRI * HTTP/2.0";
 /// version we're dealing with.
 const DEFAULT_HTTP_VERSION_DETECTION_TIMEOUT: Duration = Duration::from_secs(10);
 
+#[derive(Debug)]
+pub(crate) struct HttpHeaderFilter(Regex);
+
+#[derive(Debug)]
+pub(crate) struct HttpPathFilter(Regex);
+
+/// Current supported filtering criterias.
+#[enum_dispatch]
+#[derive(Debug)]
+pub(crate) enum HttpFilter {
+    /// Header based filter.
+    Header(HttpHeaderFilter),
+    /// Path based filter
+    Path(HttpPathFilter),
+}
+
+impl HttpFilter {
+    pub fn new_header_filter(header: Regex) -> Self {
+        Self::Header(HttpHeaderFilter(header))
+    }
+    #[allow(dead_code)]
+    pub fn new_path_filter(path: Regex) -> Self {
+        Self::Path(HttpPathFilter(path))
+    }
+}
+
+/// Struct to hold cache/hold/reuse the attributes that a [`Request`] can be matched upon
+pub(crate) struct RequestMatchCandidate<'a> {
+    /// The original request
+    request: &'a Request<Incoming>,
+    /// Headers in `k: v` format.
+    headers: OnceCell<Vec<String>>,
```

Comment:
Not understanding why we're using `OnceCell` here?

Can't we just initialize an empty `Vec`?

Reply:
The idea is to fill this lazily, only when needed. If the only filters that exist are path based, we wouldn't need to even initialize it.

Reply:
hmm I get it. I'm not a fan of lazy initialization, but think this is too minor to require arguing against.

I can see that adds more meaning, just looks a bit weird (at least initially).